ompi/mca/common/ompio/common_ompio_buffer.c

```diff
@@ -67,13 +62,15 @@ static void* mca_common_ompio_buffer_alloc_seg ( void*ctx, size_t *size )
 {
     char *buf=NULL;
     size_t realsize, numpages;
+    int tmp = -1;
+    uint64_t flags = 0;
 
     numpages = (*size + mca_common_ompio_pagesize -1 )/mca_common_ompio_pagesize;
     realsize = numpages * mca_common_ompio_pagesize;
 
     buf = malloc ( realsize);
 
-    if (NULL != buf && NULL != opal_accelerator.host_register) {
+    if (NULL != buf && 0 == opal_accelerator.check_addr(buf, &tmp, &flags)) {
```

Comment:
I don't know if this check is necessary. Can you trace the code and see if this will ever be a nonhost buffer?

Reply:
OMPIO_PREPARE_BUF() -> mca_common_ompio_alloc_buf() -> mca_common_ompio_buffer_alloc_init() -> mca_common_ompio_buffer_alloc_seg() / mca_common_ompio_buffer_free_seg()

Reply:
I think we can get in this path without checking if it was a GPU buffer.
```
int mca_common_ompio_file_write_all (ompio_file_t *fh,
                                     const void *buf,
                                     int count,
                                     struct ompi_datatype_t *datatype,
                                     ompi_status_public_t *status)
{
    int ret = OMPI_SUCCESS;
    
    if ( !( fh->f_flags & OMPIO_DATAREP_NATIVE ) &&
         !(datatype == &ompi_mpi_byte.dt  ||
           datatype == &ompi_mpi_char.dt   )) {
        /* No need to check for GPU buffer for collective I/O.
           Most algorithms first copy data to aggregators, and send/recv
           to/from GPU buffers works if ompi was compiled was GPU support.
           
           If the individual fcoll component is used: there are no aggregators 
           in that concept. However, since they call common_ompio_file_write, 
           device buffers are handled by that routine.

           Thus, we only check for
           1. Datarepresentation is anything other than 'native' and
           2. datatype is not byte or char (i.e it does require some actual
              work to be done e.g. for external32.
        */
        size_t pos=0, max_data=0;
        char *tbuf=NULL;
        opal_convertor_t convertor;
        struct iovec *decoded_iov = NULL;
        uint32_t iov_count = 0;
        
        OMPIO_PREPARE_BUF(fh,buf,count,datatype,tbuf,&convertor,max_data,decoded_iov,iov_count);    
```

Reply:
The other two cases seem to be guarded by:
```
   int is_gpu, is_managed;
    mca_common_ompio_check_gpu_buf ( fh, buf, &is_gpu, &is_managed);
    if ( is_gpu && !is_managed ) {
        need_to_copy = true;
    }
```

Reply:
same with READ.

Reply:
I think the check needs to stay.